packages/blockchain/src/consensus/clique.ts

```diff
@@ -46,7 +46,8 @@ type CliqueLatestBlockSigners = CliqueBlockSigner[]
  * This class encapsulates Clique-related consensus functionality when used with the Blockchain class.
  */
 export class CliqueConsensus implements Consensus {
-  blockchain: Blockchain
+  blockchain: Blockchain | undefined
```

Comment:
Why can this be `undefined`? Would it not be cleaner to mark `setup` as static, and then pass `blockchain` in as an argument in the (`private`) `constructor`? It seems that `blockchain` is a mandatory field?

Reply:
The goal was to not require a `blockchain` when instantiating the consensus object. Otherwise, we get a circular dependency where we want to allow users to define a custom consensus algorithm and pass that into the blockchain constructor. This was just a hack to get us there while also keeping the blockchain as an internal property of the consensus object. The alternative was to remove Blockchain from the internal properties and then require it be passed in as a parameter in every consensus method that needs the blockchain context.  I think either works but the alternative here was clunkier in practice when I implemented it

Reply:
Ah right I get your point. I need to give this some thought :thinking: